timemachine/fe/single_topology.py

```diff
@@ -0,0 +1,784 @@
+    canonical_bonds = dict()
+    for idxs, params in zip(bond_idxs, bond_params):
+        canonical_bonds[dummy.canonicalize_bond(idxs)] = params
+
+    jk = dummy.canonicalize_bond((j, k))
+
+    if jk not in canonical_bonds:
+        return False
+    elif canonical_bonds[jk][0] < 50.0:
+        return False
+    elif canonical_bonds[jk][1] < 0.05:
+        return False
+
+    return True
+
+
+def check_angle_stability(j, k, l, angle_idxs, angle_params):
+
+    canonical_angles = dict()
+    for idxs, params in zip(angle_idxs, angle_params):
+        canonical_angles[dummy.canonicalize_bond(idxs)] = params
+
+    jkl = dummy.canonicalize_bond((j, k, l))
+
+    if jkl not in canonical_angles:
+        return False
+    elif canonical_angles[jkl][0] < 50.0:
+        return False
+    elif abs(canonical_angles[jkl][1] - 0) < 0.05:
+        return False
+    elif abs(canonical_angles[jkl][1] - 3.1415) < 0.05:
+        return False
+
+    return True
+
+
+def check_bond_angle_stability(j, k, l, bond_idxs, bond_params, angle_idxs, angle_params):
+    jk_stable = check_bond_stability(j, k, bond_idxs, bond_params)
+    kl_stable = check_bond_stability(k, l, bond_idxs, bond_params)
+    jkl_stable = check_angle_stability(j, k, l, angle_idxs, angle_params)
+    return jk_stable and kl_stable and jkl_stable
+
+
+def find_attached_dummy_atoms(dg, bond_idxs, anchor):
+    attached_dummy_atoms = []
+    for a, b in bond_idxs:
+        if a == anchor and b in dg:
+            attached_dummy_atoms.append(b)
+        elif b == anchor and a in dg:
+            attached_dummy_atoms.append(a)
+    return attached_dummy_atoms
+
+
+def find_junction_bonds(anchor, bond_idxs):
+    jbs = set()
+    for i, j in bond_idxs:
+        assert i != j
+        if i == anchor or j == anchor:
+            jbs.add(dummy.canonicalize_bond((i, j)))
+    return jbs
+
+
+def enumerate_anchor_groups(anchor_idx, bond_idxs, core_idxs):
+
+    # enumerate all 1 and 2 neighbor anchor atoms to form valid anchor groups.
+
+    assert anchor_idx in core_idxs
+    assert anchor_idx in [x[0] for x in bond_idxs] or anchor_idx in [x[1] for x in bond_idxs]
+
+    nbs_1 = set()
+    nbs_2 = set()
+    for src, dst in bond_idxs:
+        if src == anchor_idx and dst in core_idxs:
+            nbs_1.add(dst)
+        elif dst == anchor_idx and src in core_idxs:
+            nbs_1.add(src)
+
+    nbs_2 = set()  # ordered tuple!
+    for atom in nbs_1:
+        for src, dst in bond_idxs:
+            if src == atom and dst in core_idxs and dst != anchor_idx:
+                nbs_2.add((atom, dst))
+            elif dst == atom and src in core_idxs and src != anchor_idx:
+                nbs_2.add((atom, src))
+
+    return nbs_1, nbs_2
+
+
+def setup_orientational_restraints(ff, mol_a, mol_b, core, dg, anchor):
+    """
+    Add restraints between dummy atoms in a dummy_group and core atoms.
+
+    mol_a's parameters are mainly used to check for numerical stability.
+    """
```

Comment:
> I'm having a hard time reviewing this function due to length and complexity -- can it be broken up into smaller pieces? (Currently ~400 lines.)

I agree there's a lot of cruft (but necessary). But I don't think the verbosity is the difficult part, but rather the varying responsibilities of this function:

1) Setting up the end-state forcefield parameters (independent of dummy atoms)
2) Detecting the geometry of (core and full) and assigning class-specific restraints
3) Verifying numerical stability of the restraints

> can asserting bond_angle_stability be done in a loop all at once at the end, rather than separately for each (anchor_core_geometry, anchor_full_geometry) branch?

Unfortunately this is needed - and specifically for ring-opening and closing:

1) If core_geometry == G1_TERMINAL: there is no way to recover from an unstable core-anchor bond. It usually implies a molecule is being disconnected, which should just assert
2) If core_geometry == G2_KINK or G3_PYRAMIDAL: some choices of core-anchor maybe okay, some are not. This behavior is specific to the geometry class. G2_KINKs have 2 choices of bonds to build cross-product restraints, and G3_PYRAMIDAL have 3 choices of bonds. But not all classes need cross-product based restraints and some eg. (centroid restraints in core_geometry==G2_KINK, full_geometry==G3_PYRAMIDAL) need *both* bonds to be present.